port invocation API to fastapi

martenson opened the pull request:
This PR refactors chunks of the workflow API to fastapi and pydantic.

It introduces one new API route alias for workflow invocation step details: `GET "/api/invocations/steps/{step_id}"`. This avoid forcing the consumer to supple `workflow_id` and `invocation_id` that are both ignored by implementation.


## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

martenson committed:
port invocation API to fastapi

martenson committed:
add the PUT request's body to schema

martenson commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -859,180 +855,6 @@ def _create_from_store(
-        PUT /api/workflows/{workflow_id}/invocations/{invocation_id}/steps/{step_id}
-        PUT /api/invocations/{invocation_id}/steps/{step_id}
-
-        Update state of running workflow step invocation - still very nebulous
```
@jmchilton is this less nebulous now?  IUUC currently this takes the `action` bool param and either unpauses or not -- is there more to it? I'd like to improve the documentation here.

davelopez commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1426,6 +1245,275 @@ def write_store(
+        trans: ProvidesUserContext = DependsOnTrans,
+        invocation_id: DecodedDatabaseIdField = InvocationIDPathParam,
+        step_details: Optional[bool] = False,
+        legacy_job_state: Optional[bool] = False,
```
Those query parameters can be annotated a bit more accurately and you can always create an alias for the annotation to reuse in all the other routes.

Recent versions of FastAPI recommend using [Annotated](https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#__tabbed_7_3) instead of the way we used to use the type hint + `Query()` as the default value. We probably should use this new annotation in all new route migrations and models.

martenson replied:
done, thanks!

davelopez commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1426,6 +1245,275 @@ def write_store(
+        step_details: Optional[bool] = False,
+        legacy_job_state: Optional[bool] = False,
+    ):
+        """A wrapper for multiple API endpoints providing the same logic."""
```
This looks like internal developer documentation, but it will be rendered in the interactive API docs at `api/docs` when you expand these routes as the "detailed description" of the route.

I suggest changing it to something more elaborate describing what the endpoint does or just removing it because it should be pretty obvious from the code for a developer and it can be a bit confusing for an API user reading through the interactive docs.

martenson replied:
done, thanks!

davelopez reviewed (commented):
Great start! those are some of the important routes!

I made some minor improvement suggestions in the inline comments but one of the "crucial" parts of porting routes to FastAPI is to model the responses so we can have nice types in the auto-generated client schema. Right now we will only have `Record<string, never>` for all the responses.

If you are planning to add them in a follow-up I guess it's fine, but I'm afraid those routes can only be considered fully ported when they have the response annotations :sweat_smile:

martenson committed:
fix copypasta

martenson committed:
imprrove and refactor models for workflow invocation API

- refactor parts of invocation schema into dedicated module
- define model for workflow invocation step
- improve sedcription of some invocation API params
- improve documentation

martenson committed:
update client schema

martenson committed:
fix errors mypy found

martenson committed:
prevent black reformatting

davelopez commented on lib/galaxy/schema/invocation.py:
```diff
@@ -214,3 +231,114 @@ class InvocationMessageResponseModel(BaseModel):
+        description="The source model of the output.",
+        const=True,
+        mark_required_in_schema=True,
+    )
```
The type of `src` should be `INVOCATION_STEP_OUTPUT_SRC` instead of `str` that way it would be considered as a constant, so you probably don't need `const=true` and `mark_required_in_schema=True`.

davelopez commented on lib/galaxy/schema/invocation.py:
```diff
@@ -214,3 +231,114 @@ class InvocationMessageResponseModel(BaseModel):
+        const=True,
+        mark_required_in_schema=True,
+    )
+    id: DecodedDatabaseIdField = Field(
```
I know at some point I tried to make all encoded IDs be of type `DecodedDatabaseIdField`, but we decided that it was more explicit to use `DecodedDatabaseIdField` for parameters and body payloads (you will get the decoded value automatically in your route handler), and `EncodedDatabaseIdField` for responses (it will ensure you have an encoded value) so it was more explicit.

So in summary, is recommended to use `DecodedDatabaseIdField` for any parameter and body payload model and `EncodedDatabaseIdField` for any ID in a response model.

davelopez commented on lib/galaxy/schema/invocation.py:
```diff
@@ -214,3 +231,114 @@ class InvocationMessageResponseModel(BaseModel):
+        description="Universal unique identifier of the workflow step.",
+    )
+    outputs: Dict[str, InvocationStepOutput] = Field(
+        {},
```
This default value might be correct (I haven't checked) but it is important to note that what makes a field `optional` in FastAPI is not using `Optional[whatever]` in the type hint (this just helps the type checker), but providing a default value.

Again, this might be correct in this particular model, but I was confused in the past by this fact so I wanted to clarify here.

The good news is that now we can easily check this by looking at the generated client schema:
```ts
/**
* Outputs
* @description The outputs of the workflow invocation step.
* @default {}
*/
outputs?: {
  [key: string]: components["schemas"]["InvocationStepOutput"] | undefined;
};
```

Observe that `outputs` can be `undefined`.

martenson replied:
The explicit `optional` vs (non)present default value is surely a bit confusing, but I am not sure it applies here. IIUC the API will return an empty dict if there are no outputs -- see https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/model/__init__.py#L8361

Hence `{}` is the correct default return value?

davelopez commented on lib/galaxy/schema/invocation.py:
```diff
@@ -214,3 +231,114 @@ class InvocationMessageResponseModel(BaseModel):
+
+
+class InvocationUpdatePayload(Model):
+    action: bool = InvocationStepAction
```
I would rename `InvocationStepAction` to `InvocationStepActionField` (or something similar) to make clear that is a Field definition otherwise it can be confused with some real "action". I know... naming is the hardest part :laughing:

davelopez reviewed (commented):
Definitely a great step for invocations! :rocket: :laughing: 

Some comments below, mostly to clarify things I had fought in the past.

martenson committed:
define invocation report response model

and add suggestions from the code review, thanks @davelopez

martenson committed:
add typing to optional fields

heisner-tillman committed:
Regenerate client schema after rebase

heisner-tillman committed:
Add pydantic model for show_invocation operation and adjust fields of model InvocationStep

heisner-tillman committed:
Add pydantic to return of show_invocation operation

martenson committed:
drop the comments

history is in the commits

heisner-tillman committed:
Regenerate client schema after rebase

heisner-tillman commented on lib/galaxy/schema/invocation.py:
```diff
@@ -218,3 +250,247 @@ class InvocationMessageResponseModel(BaseModel):
+    model_class: schema.INVOCATION_MODEL_CLASS = ModelClassField(schema.INVOCATION_MODEL_CLASS)
+    steps: List[InvocationStep] = Field(
+        default=Required, title="Steps", description="Steps of the workflow invocation."
+    )
```
Could you help me on this? @davelopez

davelopez replied:
> TODO Using the UUID4 type fails here... why?

What is the failure that you get when using UUID4 from pydantic?

> TODO - I do not know what this does and only copied it. Is this right and what does it do?

`model_class` just has to have the "literal" name of the class, the ModelClassField is simply a way of avoiding repeating the pydantic.Field information (title, description, etc.) that is similar to all those literals.

heisner-tillman replied:
> What is the failure that you get when using UUID4 from pydantic?

```
pydantic.error_wrappers.ValidationError: 1 validation error for ShowInvocationResponse
uuid
  uuid version 4 expected (type=value_error.uuid.version; required_version=4)
```

davelopez replied:
Can you debug the test that is failing to check the actual value of the `uuid` in the response? could it be None?

heisner-tillman replied:
It appears to be a UUIDv1. 
As it look like this: `xxxxxxxx-xxxx-1xxx-9xxx-xxxxxxxxxxxx`

davelopez replied:
That seems odd, If it is really a v1 I suggest changing the type then to `Union[UUID4, UUID1]` (or whatever) and add a note about it

heisner-tillman committed:
Adjust InvocationStep model to be more flexibel

heisner-tillman committed:
Adjust parameter_value of InvocationInputParameter model to support different types.

heisner-tillman committed:
Rename and add pydantic model to show_invocation and cancel_invocation operations

heisner-tillman committed:
Regenerate client schema

heisner-tillman committed:
Create pydantic model for *_jobs_summary operations

heisner-tillman committed:
Add models to *_jobs_summary operations return

heisner-tillman committed:
Resolve typing error as enum class got moved

heisner-tillman committed:
Regenerate client schema after rebase

heisner-tillman commented on lib/galaxy/model/__init__.py:
```diff
@@ -8260,7 +8257,7 @@ def active(self):
         return self.state in [states.NEW, states.READY]
 
-    def set_state(self, state: "WorkflowInvocation.states"):
+    def set_state(self, state: "InvocationState"):
```
The original enum class `states`(line 8217) got turned into an attribute and is now an instance of the seperate enum class `InvocationState`, because of this pydantic complained about typing for the set_state method. 
Is it right to change the input of the method like this?

davelopez replied:
I guess so, but you can remove the double quotes, `InvocationState` is imported in this file.
```suggestion
    def set_state(self, state: InvocationState):
```

heisner-tillman committed:
Change TODO

heisner-tillman commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -1428,6 +1276,292 @@ def write_store(
+        """An alias for `GET /api/invocations/{invocation_id}/jobs_summary`. `workflow_id` is ignored."""
+        return self.invocation_jobs_summary(trans, invocation_id)
+
+    # Should I even create models for those as they will be removed?
```
This comment is in regards to the operations export_invocation_bco, download_invocation_bco and _deprecated_generate_bco. See lines 1564 - 1648

davelopez replied:
No, you don't need to, those endpoints will be removed soon™

heisner-tillman commented on lib/galaxy/schema/invocation.py:
```diff
@@ -218,3 +251,278 @@ class InvocationMessageResponseModel(BaseModel):
+        default=Required,
+        title="Action",
+        description="Whether to take action on the invocation step.",
+    )
```
Any suggestions?

davelopez replied:
> Should I define a new enum class for this or expand the existing one?

Not an expert here, maybe someone else can give you a better answer, but I would use the existing `InvocationStepState` and expand as necessary, I have a feeling that this is/should be probably just an alias for `JobState`... or a subset of it. Also, check `InvocationJobState`... I think this should be just replaced by `JobState`?

> TODO I would like to use InvocationStepActionField here to avoid code duplication, but it is not allowed to be None. So is this ok?

If it is for reusing the Field definition, simply remove the `:bool` from the alias:
```
InvocationStepActionField = Field(
    title="Action",
    description="Whether to take action on the invocation step.",
)
```
Then you can specify the optional or not in the actual type hint and reuse the same InvocationStepActionField.

heisner-tillman commented on lib/galaxy/schema/invocation.py:
```diff
@@ -218,3 +251,278 @@ class InvocationMessageResponseModel(BaseModel):
+    # TODO - Is it ok to set parameter value to Any here? It was set to str,
+    # but that seems wrong, as it can also be an int - see test:
+    # TestWorkflowsApi::test_run_with_int_parameter
+    parameter_value: Any = Field(default=Required, title="Parameter value", description="Value of the input parameter.")
```
Is there a better type for parameter value here or is `Any` fine?

davelopez replied:
I guess you can leave it as `Any` until we narrow down all the valid types here.

heisner-tillman committed:
Remove unnecessary quotes in state type

heisner-tillman committed:
Refine pydantic models

heisner-tillman committed:
Regenerate client schema

davelopez commented on lib/galaxy/schema/invocation.py:
```diff
@@ -218,3 +253,272 @@ class InvocationMessageResponseModel(BaseModel):
+    )
+
+
+class WorkflowInvocationResponse(BaseModel):
```
```suggestion
class WorkflowInvocationResponse(Model):
```

In general, you want to make your models inherit from `schema.Model` and not directly from Pydantic `BaseModel`, the reason is, that our custom `Model` has some special features, for example, if you use `ModelClassField` it will only work for our custom base model, hence the OpenAPI linting error you are getting.

https://github.com/galaxyproject/galaxy/blob/f10614ee48532a10f3aa353ab40b409647dfc763/lib/galaxy/schema/schema.py#L280

davelopez commented on lib/galaxy/schema/invocation.py:
```diff
@@ -218,3 +253,272 @@ class InvocationMessageResponseModel(BaseModel):
+    messages: Optional[Any] = Field(default=None, title="Message", description="Message of the workflow invocation.")
+
+
+class InvocationJobsSummaryBaseModel(BaseModel):
```
```suggestion
class InvocationJobsSummaryBaseModel(Model):
```

heisner-tillman committed:
Change the inherited class from BaseModel to Model in some pydantic models

heisner-tillman committed:
Refine pydantic models

heisner-tillman committed:
Regenerate client schema

heisner-tillman committed:
Set model type of InvocationsStepJobsResponse to Any as temporary fix

heisner-tillman committed:
Regenerate client schema

heisner-tillman committed:
Add method to get pdf report of a workflow to WorkflowPopulator

heisner-tillman committed:
Add TODO and make id optional

heisner-tillman committed:
Add test for report_pdf operations of InvocationAPI

heisner-tillman committed:
Set fields to be fully optional

heisner-tillman committed:
Regenerate the client schema

mvdbeek reviewed (commented):
This is looking great, do you want to put it into review ?

heisner-tillman committed:
Regenerate the client schema

mvdbeek commented on lib/galaxy/webapps/galaxy/services/invocations.py:
```diff
@@ -146,6 +162,39 @@ def index(
+        for job_source_type, job_source_id, _ in invocation_job_source_iter(trans.sa_session, invocation_id):
+            ids.append(job_source_id)
+            types.append(job_source_type)
+        return [self.security.encode_all_ids(s) for s in fetch_job_states(trans.sa_session, ids, types)]
```
The model should take care of the id encoding/decoding

davelopez commented on lib/galaxy/schema/invocation.py:
```diff
@@ -218,3 +253,289 @@ class InvocationMessageResponseModel(BaseModel):
+
+
+class InvocationIOBase(Model):
+    # TODO - resolve
+    # the tests in test/integration/test_workflow_tasks.py ,
+    # between line 42 and 56 fail, if this id is not allowed be None
+    # They all fail, when trying to populate the response of the show_invocation operation
+    # Is it intended to allow None here?
```
Is this TODO still relevant?

davelopez commented on lib/galaxy/schema/invocation.py:
```diff
@@ -218,3 +253,289 @@ class InvocationMessageResponseModel(BaseModel):
+class InvocationStepJobsResponse(InvocationJobsSummaryBaseModel):
+    # TODO - this can also be Job or ImplicitCollectionJobs - how to handle this?
+    # model: schema.INVOCATION_STEP_MODEL_CLASS = ModelClassField(schema.INVOCATION_STEP_MODEL_CLASS)
+    model: Any
```
You usually resolve this using `Union` instead of just one model, In is that what you need?

davelopez commented on test/integration/test_workflow_tasks.py:
```diff
@@ -167,3 +168,39 @@ def _assert_one_invocation_created_and_get_details(self, response: Any) -> Dict[
+            invocation_id = summary.invocation_id
+            report_pdf = self.workflow_populator.workflow_report_pdf(workflow_id, invocation_id)
+            assert report_pdf.headers["content-type"] == "application/pdf"
+            # assert report_pdf.headers["content-disposition"] == 'attachment;'
```
```suggestion
```

davelopez reviewed (commented):
Looking good! Although the selenium test failure might be related.

heisner-tillman committed:
Remove comments

heisner-tillman committed:
Add model_class literal

heisner-tillman committed:
Create pydantic models to specify model class in the return of invocation_step_jobs_summary operations

heisner-tillman committed:
Refine return of invocation_step_jobs_summary operations

heisner-tillman committed:
Adjust type of unused path parameter for invocation_step operations

heisner-tillman committed:
Regenerate client schema

davelopez reviewed (approved):
Looks good to me!

Thank you so much @martenson and @heisner-tillman!

mvdbeek commented on lib/galaxy/schema/invocation.py:
```diff
@@ -218,3 +257,295 @@ class InvocationMessageResponseModel(BaseModel):
+    SCHEDULED = "scheduled"  # Workflow invocation step has been scheduled.
+    # CANCELLED = 'cancelled',  TODO: implement and expose
+    # FAILED = 'failed',  TODO: implement and expose
+    OK = "ok"  # Workflow invocation step has completed successfully - TODO: is this a correct description?
```
I don't think that ever happens ?

```
galaxy_main=> select count(*) from workflow_invocation_step where state='ok';
 count
-------
     0
(1 row)

galaxy_main=>
```

mvdbeek commented on lib/galaxy/schema/invocation.py:
```diff
@@ -218,3 +257,295 @@ class InvocationMessageResponseModel(BaseModel):
+        description="Format of the invocation report.",
+    )
+    markdown: Optional[str] = Field(
+        default="",
```
```suggestion
        default=None,
```

mvdbeek commented on lib/galaxy/schema/invocation.py:
```diff
@@ -218,3 +257,295 @@ class InvocationMessageResponseModel(BaseModel):
+        description="Raw galaxy-flavored markdown contents of the report.",
+    )
+    invocation_markdown: Optional[str] = Field(
+        default="",
```
```suggestion
        default=None,
```

mvdbeek commented on lib/galaxy/schema/invocation.py:
```diff
@@ -218,3 +257,295 @@ class InvocationMessageResponseModel(BaseModel):
+    generate_time: Optional[str] = schema.GenerateTimeField
+    generate_version: Optional[str] = schema.GenerateVersionField
+
+    class Config:
+        # Galaxy Report/Page response can contain many extra_rendering_data
+        # Allow any other extra fields
+        extra = Extra.allow
```
Only a fairly limited amount of extra keys are possible, `errors`, `generate_time`, `generate_version`, `history_datasets`, `workflows`, `history_dataset_collections`, `jobs`, `histories` and `invocations` afaict. We should just list them explicitly, otherwise we'll never catch misspelled attributes in the client.

mvdbeek reviewed (approved):
This is awesome, thanks a lot @heisner-tillman and @martenson!
